### ISSUE TITLE
bugprone-sizeof-expression behavior much more aggressive after 15f3cd6bfc6

### ISSUE THREAD

**Issue by tauchris** [llvm/llvm-project#57167]
After commit 15f3cd6bfc6, bugprone-sizeof-expression is flagging many new cases (which were not covered by existing LIT test).  Whether or not the new warnings are considered correct, or even desirable, is perhaps subjective; but I do not think the new behavior is reasonable.

I prepared this reproducer to illustrate behavior of relevant cases before the upstream ElaboratedType sugaring change (labeled "Before 15f3cd6bfc6"), and after that upstream change (labeled "After 15f3cd6bfc6").  Note that this code can be compiled as either "C" or "C++", and that was done purposefully to illustrate the behavior differences in the checker that existed before the ElaboratedType sugaring commit.

```cpp
     1       #ifdef __cplusplus
     2       #define STRKWD
     3       #else
     4       #define STRKWD struct
     5       #endif
     6       
     7       int Test5() {
     8         typedef int Array10[10];
     9       
    10         struct MyStruct {
    11           Array10 arr;
    12           Array10* ptr;
    13         };
    14       
    15         typedef struct TypedefStruct {
    16           Array10 arr;
    17           Array10* ptr;
    18         } TypedefStruct;
    19       
    20         typedef const STRKWD MyStruct TMyStruct;
    21         typedef const STRKWD MyStruct *PMyStruct;
    22         typedef TMyStruct *PMyStruct2;
    23         typedef const TypedefStruct *PTTStruct;
    24       
    25         STRKWD MyStruct S;
    26         TypedefStruct TS;
    27         PMyStruct PS;
    28         PMyStruct2 PS2;
    29         Array10 A10;
    30         PTTStruct PTTS;
    31       
    32         int sum = 0;
    33                                                     // Before 15f3cd6bfc6   After 15f3cd6bfc6
    34         sum += sizeof(&S);                          // (C++:WARN C:WARN)    (C++/C: WARN)
    35         sum += sizeof(__typeof(&S));                // (C++:WARN C:    )    (C++/C: WARN)
    36         sum += sizeof(&TS);                         // (C++:WARN C:WARN)    (C++/C: WARN)
    37         sum += sizeof(__typeof(&TS));               // (C++:     C:    )    (C++/C: WARN)
    38         sum += sizeof(STRKWD MyStruct*);            // (C++:WARN C:    )    (C++/C: WARN)
    39         sum += sizeof(__typeof(STRKWD MyStruct*));  // (C++:WARN C:    )    (C++/C: WARN)
    40         sum += sizeof(TypedefStruct*);              // (C++:     C:    )    (C++/C: WARN)
    41         sum += sizeof(__typeof(TypedefStruct*));    // (C++:     C:    )    (C++/C: WARN)
    42         sum += sizeof(PTTS);                        // (C++:WARN C:WARN)    (C++/C: WARN)
    43         sum += sizeof(PMyStruct);                   // (C++:WARN C:    )    (C++/C: WARN)
    44         sum += sizeof(PMyStruct2);                  // (C++:     C:    )    (C++/C: WARN)
    45         sum += sizeof(PS);                          // (C++:WARN C:WARN)    (C++/C: WARN)
    46         sum += sizeof(PS2);                         // (C++:WARN C:WARN)    (C++/C: WARN)
    47         sum += sizeof(&A10);                        // (C++:WARN C:WARN)    (C++/C: WARN)
    48       
    49         return sum;
    50       } 
```

The code above can be saved as sizeof-expression.c, and then analyzed as C or C++ code, using the following commands, respectively:
```console
> clang-tidy --checks=-*,bugprone-sizeof-expression sizeof-expression.c -- -x c
> clang-tidy --checks=-*,bugprone-sizeof-expression sizeof-expression.c -- -x c++
```
Notable observations are that the upstream ElaboratedType change made the behavior of the checker consistent between C and C++, where before it was not – however, it also introduced the following differences:

1. Prior to the change, `struct` declarations that are nested in `typedef` declarations produced warnings (in C++ only) when using the `struct` tag in the `sizeof()` expression, but did not produce warnings when using the `typedef` name.  After the change, warnings are issued in both C++ and C, and are issued whether the struct tag name or the `typedef` name is used.  (See lines 38 and 40.)
2. Prior to the change, use of the `__typeof()` builtin would avoid warnings in cases where warnings might otherwise be generated (see line 37, in both C++ and C).  After the change, use of `__typeof()` has no effect.

In my opinion, taking `sizeof(X *)` is a common and natural thing to do, and I see it frequently in our code base.  It's obviously useful and necessary in cases where there are different types of pointers (different address spaces, or near vs far, e.g.), where different pointers may have different sizes.  The fact that `X` happens to be a `struct`, or a `typedef` to a `struct`, should not matter.   While the new behavior is at least consistent between C and C++, and consistent regardless of declaration style (whether you are using a struct tag name or a `typedef` name in the `sizeof()` expression), it simply warns in too many cases now.  I would suggest that the checker has been too aggressive for some time, and that the ElaboratedType change simply exposed that overzealous matching, which was previously hidden due to the ambiguity in desugaring that used to exist before that change.

In any case, the fact that `__typeof()` is no longer a usable way to suppress warnings from this checker is a bug (I assume in the checker itself, though I'm not sure).

**Comment by mizvekov** [comment]
Thanks for reporting this!

I am not sure that the fact `__typeof` suppressed the warning was intentional, this is not covered in the tests.

I don't think it would be unreasonable if it did, but perhaps it would be unreasonable if it was the only way to do so.

What about pure ISO C programs, how would they suppress it?

Should we also suppress with `decltype`?

**Comment by tauchris** [comment]
I've posted a candidate fix as https://reviews.llvm.org/D131926.

@mizvekov , TBH, I'm not sure about __typeof, either.  It's not clear to me that this is the intent of this keyword, but it certainly seems relevant.  As you say, it was not covered in the existing LIT test, but then, neither was C, and that behavior clearly differed substantially.  I've included a cross-language test that exercises these cases in the differential (basically, it's the reproducers I show here, with FileCheck directives).

Regarding ways of suppressing this warning, there is always the option of adding "// NOLINT(bugprone-sizeof-expression)" to the code.  I would not want to put users in a position of having to annotate every place where their code triggers a warning from the updated version of the checker, though, if the existing code makes frequent and legitimate use of "sizeof(T*)".  That's labor-intensive and ugly.  It's better to make the checker not warn for that case, IMO.  Or perhaps, a default-fase option to enable this warning behavior...  That's certainly another way to go...

decltype is a C++-only keyword.  I don't have a clear opinion about whether this checker should explicitly treat decltype differently, but it's irrelevant for C code, which is where I've noticed the biggest behavior differences.

**Comment by tauchris** [comment]
D131926 addressed this bug.
